File: core/hash/knuth_hash.cs
/**
 * @file knuth_hash.cs
 * @brief C# implementation of Knuth multiplicative hash algorithm
 *
 * This implementation provides identical hash values to the C++ version
 * (knuth_hash.cpp) for the same input strings. This is useful for
 * cross-platform scenarios where both C++ and C# code need to generate
 * consistent hash values.
 *
 * Use case: Converting file paths to integer primary keys in a database
 * that is accessed by both C++ and C# applications.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace KnuthHashExample
{
    class Program
    {
        /// <summary>
        /// Calculate Knuth multiplicative hash for a string
        /// </summary>
        /// <param name="read">Input string</param>
        /// <returns>64-bit unsigned hash value</returns>
        /// <remarks>
        /// Algorithm:
        /// - Initial value: 3074457345618258791 (large prime)
        /// - For each character: hash = (hash + char) * 3074457345618258799
        /// </remarks>
        static UInt64 CalculateHash(string read)
        {
            UInt64 hashedValue = 3074457345618258791ul;
            for (int i = 0; i < read.Length; i++)
            {
                hashedValue += read[i];
                hashedValue *= 3074457345618258799ul;
            }
            return hashedValue;
        }

        static void Main(string[] args)
        {
            // Example: Hash a file path (case-insensitive)
            string str = @"C:\Users\user\Desktop\test_file.png";
            str = str.ToLower();

            Console.WriteLine(str);
            Console.WriteLine(CalculateHash(str));
        }
    }
}

File: core/hash/knuth_hash.cpp
/**
 * @file knuth_hash.cpp
 * @brief Knuth multiplicative hash algorithm for string to uint64 conversion
 *
 * This implementation provides a consistent hash value for strings across
 * C++ and C# platforms. Useful for scenarios where string-based keys need
 * to be converted to integer primary keys (e.g., file paths as database PKs).
 *
 * Key features:
 * - Uses Knuth's multiplicative hash constants
 * - Produces 64-bit unsigned integer hash values
 * - Compatible with the C# implementation (see knuth_hash.cs)
 *
 * Note: This example uses MFC's CString. For standard C++, see the README
 * for std::string or std::wstring alternatives.
 */

#include <locale>
#include <atlstr.h>

using namespace std;

/**
 * Calculate Knuth multiplicative hash for a string
 * @param read Input string (CString)
 * @return 64-bit unsigned hash value
 *
 * Algorithm:
 * - Initial value: 3074457345618258791 (large prime)
 * - For each character: hash = (hash + char) * 3074457345618258799
 */
UINT64 CalculateHash(CString read)
{
	UINT64 hashedValue = 3074457345618258791ull;
	for (int i = 0; i < read.GetLength(); i++)
	{
		hashedValue += read[i];
		hashedValue *= 3074457345618258799ull;
	}
	return hashedValue;
}

int _tmain(int argc, _TCHAR* argv[])
{
	// Set locale for proper Unicode handling
	locale::global(locale(""));

	// Example: Hash a file path (case-insensitive)
	CString str = R"(C:\Users\user\Desktop\test_file.png)";
	str.MakeLower();

	// Display the string and its hash value
	printf("%ws\n", str.GetBuffer());
	printf("%llu\n", CalculateHash(str));

	return 0;
}